parlai/agents/image_seq2seq/modules.py

```diff
@@ -219,11 +246,65 @@ def forward(
                 'set correctly.'
             )
 
-        full_enc = self.cat([context_encoded, image_encoded])
-        full_mask = self.cat([context_mask, extra_masks])
+        if self.image_combination_mode == 'add':
+            full_enc = self._add([context_encoded, image_encoded])
+            # image_encoded broadcasted along dim=1
+            full_mask = context_mask
+        elif self.image_combination_mode == 'append':
+            full_enc = self._cat([context_encoded, image_encoded])
+            full_mask = self._cat([context_mask, extra_masks])
+        elif self.image_combination_mode == 'prepend':
+            full_enc = self._cat([image_encoded, context_encoded])
+            full_mask = self._cat([extra_masks, context_mask])
+        else:
+            raise ValueError('Image combination mode not recognized!')
+
+        # In fp16 mode, either remove extra tokens or add new ones on to get to a
+        # multiple of 8
+        if full_enc.dtype == torch.half and full_mask is not None:
```

Comment:
could we move this logic out of `forward` into a separate function?

Reply:
Yes - moved